crates/ra_hir/src/expr.rs

```diff
@@ -720,10 +734,27 @@ impl ExprCollector {
                 let exprs = e.exprs().map(|expr| self.collect_expr(expr)).collect();
                 self.alloc_expr(Expr::Tuple { exprs }, syntax_ptr)
             }
+
             ast::ExprKind::ArrayExpr(e) => {
-                let exprs = e.exprs().map(|expr| self.collect_expr(expr)).collect();
-                self.alloc_expr(Expr::Array { exprs }, syntax_ptr)
+                let kind = e.kind();
+
+                match kind {
+                    ArrayExprKind::ElementList(e) => {
+                        let exprs = e.map(|expr| self.collect_expr(expr)).collect();
+                        self.alloc_expr(Expr::Array(Array::ElementList(exprs)), syntax_ptr)
+                    }
+                    ArrayExprKind::Repeat { initializer, repeat } => {
+                        // unwrap because an array repeat expr must have and init followed by a repeat
+                        let initializer = initializer.map(|expr| self.collect_expr(expr)).unwrap();
+                        let repeat = repeat.map(|expr| self.collect_expr(expr)).unwrap();
```

Comment:
Should I've used collect_expr_opt here instead ?. Because it this unwrap does fail it means that there is a problem in the parser.

Reply:
Yeah, it's better to use `_opt` here. Parser very intentionally produces "partial" syntax tree. For example, for something like `[;]` you will get an array expr where both initializer and repeat are missing. This makes analysis resilient to syntax erros.

In other words, all of the AST method return options for a reason. You don't have emit errors on them (this is done by the parser), but you must handle them gracefully.